File: tests/parsers/assignment-parser.test.ts
import { describe, it, expect } from 'vitest';
import { parseAssignments } from '../../src/parsers/assignment-parser';
import * as fs from 'fs';

describe('parseAssignments', () => {
  it('should parse assignment tile structure from real VULMS HTML', () => {
    const html = `<!DOCTYPE html>
<html>
<head><title>Assignment</title></head>
<body>
  <h3 class="m-subheader__title">CS301 - Data Structures</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment # 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0">8. Implementation of Stack</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Apr 30, 2026</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20.00</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Submitted</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0">Apr 25, 2026 03:39 PM</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0">2.58 KB</span>
</body>
</html>`;

    const { assignments, confidence } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0]).toMatchObject({
      courseCode: 'CS301',
      title: 'Assignment # 1',
      lesson: '8. Implementation of Stack',
      totalMarks: 20,
      status: 'submitted',
      fileSize: '2.58 KB',
    });
    expect(confidence.extracted).toBe(1);
    expect(confidence.confidence).toBeGreaterThan(0);
  });

  it('should parse submitted status', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Submitted</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">May 15, 2024</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('submitted');
  });

  it('should parse result_declared status', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS201 - Data Structures</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0">8</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Result Declared</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('result_declared');
    expect(assignments[0].obtainedMarks).toBe(8);
  });

  it('should parse missed status', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS301 - Data Structures</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Not Submitted</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('missed');
  });

  it('should parse pending status by default', () => {
    const futureYear = new Date().getFullYear() + 1;
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS301 - Data Structures</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Aug 01, ${futureYear}</span>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_1">Assignment 2</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_1">Aug 01, ${futureYear}</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_1">10</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_1"></span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(2);
    const emptySubmit = assignments.find(a => a.title === 'Assignment 2');
    expect(emptySubmit?.status).toBe('pending');
  });

  it('should return empty when no assignments found', () => {
    const html = '<html><body>No assignments</body></html>';
    const { assignments, confidence } = parseAssignments(html);
    expect(assignments).toEqual([]);
    expect(confidence.extracted).toBe(0);
    expect(confidence.confidence).toBeLessThan(1);
  });

  it('should parse real VULMS assignment fixture', () => {
    const html = fs.readFileSync('tests/fixtures/assignments.html', 'utf8');
    const { assignments, confidence } = parseAssignments(html);
    expect(assignments.length).toBeGreaterThan(0);
    expect(assignments[0].courseCode).toBe('CS301');
    expect(assignments[0].title).toBe('Assignment # 1');
    expect(confidence.extracted).toBeGreaterThan(0);
  });

  it('should classify submitted assignment with due date as submitted', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Submitted</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Jan 01, 2020</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0">Dec 25, 2019 10:00 AM</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('submitted');
  });

  it('should classify unsubmitted + future due date as pending', () => {
    const futureYear = new Date().getFullYear() + 1;
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Dec 01, ${futureYear}</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('pending');
  });

  it('should classify unsubmitted + past due date as missed', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Jan 01, 2020</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('missed');
  });

  it('should classify rawStatus=Expired as missed', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Expired</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Jan 01, 2020</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('missed');
  });

  it('should classify rawStatus=Result Declared as result_declared', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Result Declared</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0">15</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('result_declared');
    expect(assignments[0].obtainedMarks).toBe(15);
  });

  it('should classify unsubmitted + invalid due date as pending fallback', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">-</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('pending');
  });

  it('should classify MTH202 real fixture as missed (past due, no submission)', () => {
    const html = fs.readFileSync('debug/navigation/assignments/MTH202.html', 'utf8');
    const { assignments } = parseAssignments(html, undefined, { courseCode: 'MTH202' });
    expect(assignments.length).toBeGreaterThan(0);
    for (const a of assignments) {
      expect(a.status).toBe('missed');
    }
  });

  it('should classify submitted assignment even when due date is past', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3 class="m-subheader__title">CS101 - Intro</h3>
  <span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Submitted</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">Jan 01, 2020</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0">Dec 30, 2019 05:00 PM</span>
  <span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">10</span>
</body>
</html>`;

    const { assignments } = parseAssignments(html);
    expect(assignments).toHaveLength(1);
    expect(assignments[0].status).toBe('submitted');
    expect(assignments[0].dueDate).toBeDefined();
    expect(assignments[0].submitDate).toBeDefined();
  });
});

File: vulms_dart/test/parsers/assignment_parser_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:vulms_dart/src/parsers/assignment_parser.dart';
import 'package:vulms_dart/src/models/assignment.dart';

void main() {
  final standardFixture =
      File('test/fixtures/assignments_standard.html').readAsStringSync();
  final legacyFixture =
      File('test/fixtures/assignments_legacy.html').readAsStringSync();
  final emptyFixture =
      File('test/fixtures/assignments_empty.html').readAsStringSync();
  final emptyPageFixture =
      File('test/fixtures/empty_page.html').readAsStringSync();

  group('AssignmentParser.parse()', () {
    group('standard layout (Label3 IDs)', () {
      test('parses 2 assignments', () {
        final result = AssignmentParser.parse(standardFixture);
        expect(result, hasLength(2));
      });

      test('extracts assignment 1 details', () {
        final result = AssignmentParser.parse(standardFixture);
        final a1 = result.firstWhere((a) => a.title.contains('OOP'));
        expect(a1.courseCode, 'CS101');
        expect(a1.title, 'Assignment 1: OOP Concepts');
        expect(a1.lesson, 'Lesson 5');
        expect(a1.totalMarks, 20.0);
        expect(a1.status, AssignmentStatus.submitted);
        expect(a1.fileSize, '2.5 MB');
        expect(a1.obtainedMarks, 18.0);
      });

      test('extracts assignment 2 details', () {
        final result = AssignmentParser.parse(standardFixture);
        final a2 = result.firstWhere((a) => a.title.contains('Data'));
        expect(a2.title, 'Assignment 2: Data Structures');
        expect(a2.lesson, 'Lesson 8');
        expect(a2.totalMarks, 25.0);
        expect(a2.status, AssignmentStatus.missed);
        expect(a2.fileSize, '-');
        expect(a2.obtainedMarks, isNull);
      });
    });

    group('legacy layout (lblTitle IDs)', () {
      test('parses 1 assignment', () {
        final result = AssignmentParser.parse(legacyFixture);
        expect(result, hasLength(1));
      });

      test('extracts assignment details from legacy layout', () {
        final result = AssignmentParser.parse(legacyFixture);
        final a = result.first;
        expect(a.courseCode, 'MGT301');
        expect(a.title, 'Assignment 1: Management Principles');
        expect(a.lesson, 'Lesson 3');
        expect(a.totalMarks, 15.0);
        expect(a.status, AssignmentStatus.resultDeclared);
        expect(a.fileSize, '1.2 MB');
        expect(a.obtainedMarks, 12.0);
      });
    });

    group('empty assignments page', () {
      test('returns empty list', () {
        final result = AssignmentParser.parse(emptyFixture);
        expect(result, isEmpty);
      });

      test('returns empty list for page with no repeater', () {
        final result = AssignmentParser.parse(emptyPageFixture);
        expect(result, isEmpty);
      });
    });

    group('status derivation', () {
      test('Submitted results in submitted status', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0">Test</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Submitted</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0">14-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result.first.status, AssignmentStatus.submitted);
      });

      test('Not Submitted results in missed status when due date is past', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0">Test</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">01-Jan-2020</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0">Not Submitted</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0">-</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result.first.status, AssignmentStatus.missed);
      });

      test('Result results in resultDeclared status', () {
        final result = AssignmentParser.parse(legacyFixture);
        expect(result.first.status, AssignmentStatus.resultDeclared);
      });
    });

    group('marks parsing', () {
      test('parses marks with comma as 1500.0', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0">Test</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">1,500</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result.first.totalMarks, 1500.0);
      });

      test('dash marks return null', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0">Test</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">-</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result.first.totalMarks, isNull);
      });
    });

    group('deduplication', () {
      test('removes duplicate assignments with same title', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0">Assignment 1</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_1">Assignment 1</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_1"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_1">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_1">20</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_1"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_1"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_1"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_1"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result, hasLength(1));
      });
    });

    group('empty titles', () {
      test('skips assignments with empty titles', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<div id="MainContent_gvTileRepeaterAssignment">'
            '<span id="MainContent_gvTileRepeaterAssignment_Label3_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblPayableAmount_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblDueDate_0">15-Jan-2024</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblTotalMarks_0">20</span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblsubmitted_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblSubmitDate_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblFilesize_0"></span>'
            '<span id="MainContent_gvTileRepeaterAssignment_lblObtainedMarks_0"></span>'
            '</div></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result, isEmpty);
      });
    });

    group('course code extraction', () {
      test('extracts course code from h3.m-subheader__title', () {
        final result = AssignmentParser.parse(standardFixture);
        expect(result.first.courseCode, 'CS101');
      });

      test('extracts course code from title element', () {
        final result = AssignmentParser.parse(legacyFixture);
        expect(result.first.courseCode, 'MGT301');
      });
    });

    group('table fallback', () {
      test('falls back to table when no tile elements', () {
        const html = '<html><body>'
            '<table>'
            '<tr><th>Title</th><th>Due Date</th><th>Marks</th><th>Status</th></tr>'
            '<tr><td>Assignment 1</td><td>15-Jan-2024</td><td>20</td><td>Submitted</td></tr>'
            '</table></body></html>';
        final result = AssignmentParser.parse(html);
        expect(result, hasLength(1));
        expect(result.first.title, 'Assignment 1');
      });
    });

    test('accepts explicit courseCode parameter', () {
      final result = AssignmentParser.parse(standardFixture, courseCode: 'XX999');
      expect(result.every((a) => a.courseCode == 'XX999'), isTrue);
    });
  });
}
